File: src/features.py
"""แปลงจุด landmark ของมือให้เป็น "เวกเตอร์คุณลักษณะ" (feature vector)

นี่คือหัวใจของโปรเจค และเป็นเหตุผลที่วิธีนี้ใช้ข้อมูลน้อยกว่าการเทรนจากรูปภาพมาก

ปัญหาคือ พิกัดดิบจาก MediaPipe ขึ้นกับสิ่งที่ไม่เกี่ยวกับความหมายของท่ามือเลย:
    - มืออยู่ตรงไหนของเฟรม  (ตำแหน่ง)
    - มืออยู่ใกล้หรือไกลกล้อง (ขนาด)
เราจึงต้อง normalize ทิ้งสองอย่างนี้ออกไป เหลือไว้แค่ "รูปทรงของมือ" ล้วน ๆ:

    1. ย้ายจุดข้อมือ (landmark 0) ไปไว้ที่จุดกำเนิด  -> ตัดผลของตำแหน่ง
    2. หารด้วยระยะที่ไกลที่สุดจากข้อมือ              -> ตัดผลของขนาด

ผลลัพธ์: ทำท่าเดียวกันที่มุมไหนของจอ ใกล้หรือไกลแค่ไหน ก็ได้ตัวเลขชุดเดิม

แต่การ normalize แบบนี้มีราคาที่ต้องจ่าย: เมื่อแต่ละมือถูกเทียบกับข้อมือของตัวเอง
ความสัมพันธ์ระหว่างสองมือก็หายไปด้วย จึงต่อท้ายอีก 5 ค่าเพื่อเก็บส่วนนั้นกลับมา
(ดูหัวข้อ "ท่อนคู่มือ" ข้างล่าง) รวมเป็น 133 ค่า
"""

from __future__ import annotations

import numpy as np

NUM_LANDMARKS = 21
COORDS_PER_LANDMARK = 3          # x, y, z
PER_HAND_DIMS = NUM_LANDMARKS * COORDS_PER_LANDMARK + 1   # +1 = ธงบอกว่ามีมือข้างนี้ไหม
HAND_SLOTS = ("Left", "Right")   # ล็อกตำแหน่งในเวกเตอร์ให้คงที่เสมอ
HANDS_DIMS = PER_HAND_DIMS * len(HAND_SLOTS)              # = 128

# ---------------------------------------------------------------- ท่อนคู่มือ
# การ normalize ข้างบนเทียบกับข้อมือ "ของมือข้างนั้นเอง" ผลข้างเคียงคือความสัมพันธ์
# ระหว่างสองมือหายไปหมด: มือประกบกันกับมือแยกห่างกันจะได้ตัวเลขชุดเดียวกันเป๊ะ
# ซึ่งใช้ไม่ได้กับภาษามือไทยหลายคำที่ระยะห่างของสองมือเป็นตัวแยกความหมาย
#
# จึงต่อท้ายอีก 5 ค่าเพื่อเก็บ "ความสัมพันธ์ระหว่างสองมือ" ที่หายไป
PAIR_DIMS = 5
PAIR_START = HANDS_DIMS
FEATURE_DIMS = HANDS_DIMS + PAIR_DIMS                     # = 133

# จำนวนมิติของชุดข้อมูลรุ่นก่อนที่จะมีท่อนคู่มือ ใช้ตอนนำเข้าข้อมูลเก่า
LEGACY_FEATURE_DIMS = HANDS_DIMS


def normalize_landmarks(landmarks: np.ndarray) -> np.ndarray:
    """normalize จุด (21, 3) ของมือหนึ่งข้าง แล้วคืนเวกเตอร์แบน 63 ค่า"""
    normalized, _wrist, _scale = normalize_with_geometry(landmarks)
    return normalized


def normalize_with_geometry(landmarks: np.ndarray):
    """เหมือน normalize_landmarks แต่คืนตำแหน่งข้อมือและสเกลที่หารทิ้งไปด้วย

    สองค่าที่ปกติถูก "ทิ้ง" นี้คือสิ่งที่ท่อนคู่มือต้องใช้: ตำแหน่งข้อมือบอกว่า
    สองมืออยู่ห่างกันแค่ไหน และสเกลบอกว่ามือข้างไหนอยู่ใกล้กล้องกว่า
    """
    points = np.asarray(landmarks, dtype=np.float32).reshape(NUM_LANDMARKS, COORDS_PER_LANDMARK)

    # 1) ย้ายข้อมือไปที่จุดกำเนิด
    centered = points - points[0]

    # 2) ปรับสเกลด้วยระยะที่ไกลสุดจากข้อมือ
    scale = np.linalg.norm(centered, axis=1).max()
    if scale < 1e-6:          # กันหารด้วยศูนย์ ถ้าจุดซ้อนกันหมด
        scale = 1.0

    return (centered / scale).flatten(), points[0], scale


def build_feature_vector(hands) -> np.ndarray:
    """รวมมือที่ตรวจเจอ (0-2 ข้าง) เป็นเวกเตอร์ความยาวคงที่ 133 ค่า

    ต้องมีความยาวคงที่เพราะโมเดล machine learning รับ input ขนาดตายตัว
    ถ้าเจอมือข้างเดียว ช่องของอีกข้างจะเป็นศูนย์ทั้งหมด และธงบอกการมีอยู่ = 0

    โครงของเวกเตอร์:
        [  0..63 ]  มือ Left  : 63 ค่า normalize แล้ว + ธงบอกการมีอยู่
        [ 64..127]  มือ Right : เหมือนกัน
        [128..132]  ท่อนคู่มือ : offset x/y/z, สมดุลขนาด, ธงบอกว่าค่าใช้ได้
    """
    vector = np.zeros(FEATURE_DIMS, dtype=np.float32)
    wrists: dict[int, np.ndarray] = {}
    scales: dict[int, float] = {}

    for hand in hands:
        if hand.handedness not in HAND_SLOTS:
            continue
        slot = HAND_SLOTS.index(hand.handedness)
        start = slot * PER_HAND_DIMS
        normalized, wrist, scale = normalize_with_geometry(hand.landmarks)
        vector[start:start + PER_HAND_DIMS - 1] = normalized
        vector[start + PER_HAND_DIMS - 1] = 1.0   # ธง: มือข้างนี้ปรากฏอยู่
        wrists[slot] = wrist
        scales[slot] = scale

    if len(wrists) == len(HAND_SLOTS):
        _fill_pair_block(vector, wrists, scales)

    return vector


def _fill_pair_block(vector: np.ndarray, wrists: dict, scales: dict) -> None:
    """เติมท่อนคู่มือ — เรียกเฉพาะเมื่อเจอมือครบสองข้างเท่านั้น

    offset ถูกหารด้วยขนาดมือเฉลี่ย เพื่อให้ยังไม่ขึ้นกับระยะห่างจากกล้อง
    (ยืนไกลขึ้น ทั้ง offset และขนาดมือหดลงเท่ากัน อัตราส่วนจึงคงเดิม)

    สมดุลขนาดใช้สูตร (ขวา-ซ้าย)/(ขวา+ซ้าย) ซึ่งอยู่ในช่วง -1 ถึง 1 เสมอ
    เลือกแบบนี้แทนอัตราส่วนตรง ๆ เพราะไม่ระเบิดเมื่อตัวหารเล็ก และสลับเครื่องหมาย
    พอดีเมื่อกลับด้านซ้าย-ขวา (ดู mirror_feature_vector)
    """
    left_slot, right_slot = 0, 1
    left_scale = np.float32(scales[left_slot])
    right_scale = np.float32(scales[right_slot])

    mean_scale = (left_scale + right_scale) / np.float32(2.0)
    if mean_scale < 1e-6:
        mean_scale = np.float32(1.0)

    offset = (wrists[right_slot] - wrists[left_slot]) / mean_scale
    vector[PAIR_START:PAIR_START + 3] = offset

    total_scale = left_scale + right_scale
    if total_scale > 1e-6:
        vector[PAIR_START + 3] = (right_scale - left_scale) / total_scale

    vector[PAIR_START + 4] = 1.0   # ธง: ค่าในท่อนนี้วัดมาจริง ไม่ใช่ศูนย์เพราะไม่รู้


def build_canonical_feature_vector(hands, source_is_mirrored: bool) -> np.ndarray:
    """ประกอบเวกเตอร์ให้อยู่ใน "คอนเวนชันกลาง" ที่ทั้งคอมพิวเตอร์และมือถือใช้ร่วมกัน

    คอนเวนชันกลางคือ **ภาพดิบจากกล้อง (ยังไม่พลิกกระจก)** เหตุผลที่เลือกอันนี้
    เพราะฝั่งมือถือส่งภาพดิบเข้า MediaPipe อยู่แล้ว (พลิกกระจกด้วย CSS ตอนแสดงผล
    เท่านั้น) ฝั่งคอมพิวเตอร์จึงเป็นฝ่ายที่ต้องปรับให้ตรงกัน

    ทำไมฝั่งคอมพิวเตอร์ต้องปรับ: `src/collect.py` และ `src/translate.py` พลิกเฟรม
    ด้วย cv2.flip **ก่อน** ส่งเข้า MediaPipe (เพื่อให้พิกัดตรงกับภาพบนจอ วาดโครงมือ
    ได้ง่าย) ผลคือได้เวกเตอร์ของโลกกระจกเงา: แกน x กลับด้าน และ MediaPipe เรียก
    มือขวาจริงว่า "Left" ซึ่งเป็นภาพสะท้อนของคอนเวนชันกลางพอดี จึงแปลงกลับได้ด้วย
    mirror_feature_vector เพียงครั้งเดียว

    ถ้าไม่แปลง ข้อมูลจากสองฝั่งจะลงช่องมือสลับข้างกันและ x กลับด้าน ใช้ร่วมกันไม่ได้
    โดยไม่มี error ฟ้อง (ตรวจสอบด้วย `run parity` และ `python selftest.py`)
    """
    vector = build_feature_vector(hands)
    if source_is_mirrored:
        vector = mirror_feature_vector(vector)
    return vector


def mirror_feature_vector(vector: np.ndarray) -> np.ndarray:
    """สร้างเวกเตอร์ของภาพสะท้อนกระจก — ใช้เพิ่มข้อมูลตอนเทรน (data augmentation)

    ท่ามือส่วนใหญ่ทำมือซ้ายหรือขวาก็สื่อความหมายเดียวกัน การกลับด้านข้อมูล
    จึงทำให้ได้ตัวอย่างเพิ่มฟรี ๆ เท่าตัว และโมเดลไม่ยึดติดว่าต้องใช้มือข้างไหน

    การกลับด้านทำสองอย่างพร้อมกัน: สลับเครื่องหมายแกน x และสลับช่องซ้าย-ขวา
    """
    mirrored = np.zeros_like(vector)

    for slot in range(len(HAND_SLOTS)):
        start = slot * PER_HAND_DIMS
        coords = vector[start:start + PER_HAND_DIMS - 1].reshape(NUM_LANDMARKS, COORDS_PER_LANDMARK)
        present = vector[start + PER_HAND_DIMS - 1]

        flipped = coords.copy()
        flipped[:, 0] *= -1.0        # กลับแกน x

        other = (slot + 1) % len(HAND_SLOTS)
        dest = other * PER_HAND_DIMS
        mirrored[dest:dest + PER_HAND_DIMS - 1] = flipped.flatten()
        mirrored[dest + PER_HAND_DIMS - 1] = present

    # ท่อนคู่มือกลับด้าน "ตรงข้าม" กับท่อนของแต่ละมือ ซึ่งดูขัดความรู้สึกแต่ถูกต้อง
    #
    # offset นิยามว่า (ข้อมือขวา - ข้อมือซ้าย) การกลับด้านทำให้เกิดสองอย่างพร้อมกัน
    # คือ x ของทุกจุดสลับเครื่องหมาย **และ** ช่องซ้าย-ขวาสลับกัน ซึ่งกลับทิศของ
    # ตัวลบเอง สองอย่างนี้หักล้างกันพอดีบนแกน x จึงคงเดิม ส่วนแกน y กับ z
    # ไม่ถูกกลับจากการพลิกภาพ เหลือแต่ผลจากการสลับตัวลบ จึงสลับเครื่องหมาย
    #
    #   offset'  = ข้อมือขวาใหม่ - ข้อมือซ้ายใหม่ = M(ข้อมือซ้าย) - M(ข้อมือขวา)
    #   offset'.x = (1-ซ้าย.x) - (1-ขวา.x) = ขวา.x - ซ้าย.x =  offset.x   (คงเดิม)
    #   offset'.y = ซ้าย.y - ขวา.y                          = -offset.y   (สลับ)
    mirrored[PAIR_START] = vector[PAIR_START]
    mirrored[PAIR_START + 1] = -vector[PAIR_START + 1]
    mirrored[PAIR_START + 2] = -vector[PAIR_START + 2]
    # สมดุลขนาด: มือขวากลายเป็นมือซ้าย ตัวตั้งกับตัวลบสลับกัน จึงสลับเครื่องหมาย
    mirrored[PAIR_START + 3] = -vector[PAIR_START + 3]
    mirrored[PAIR_START + 4] = vector[PAIR_START + 4]

    return mirrored


def upgrade_legacy_vector(values: np.ndarray) -> np.ndarray:
    """ขยายเวกเตอร์รุ่นเก่า 128 ค่า ให้เป็น 133 ค่า

    ท่อนคู่มือของข้อมูลเก่ากู้กลับไม่ได้จริง ๆ เพราะตำแหน่งข้อมือและขนาดมือถูก
    หารทิ้งไปตอนเก็บ จึงเติมศูนย์แล้วปล่อยธงเป็น 0 ซึ่งหมายถึง "ไม่รู้ค่า"
    ไม่ใช่ "สองมือซ้อนทับกันพอดี" โมเดลเรียนรู้ที่จะมองข้ามท่อนนี้เมื่อธงเป็น 0
    ได้เอง เหมือนที่มันเรียนรู้จากธงบอกการมีอยู่ของแต่ละมือ

    สำหรับตัวอย่างที่มีมือข้างเดียวหรือไม่มีมือเลย ค่าที่ถูกคือศูนย์และธงเป็น 0
    อยู่แล้ว ข้อมูลกลุ่มนี้จึงถูกกู้กลับได้ครบถ้วนไม่มีอะไรเสียหาย
    """
    upgraded = np.zeros(FEATURE_DIMS, dtype=np.float32)
    upgraded[:LEGACY_FEATURE_DIMS] = np.asarray(values, dtype=np.float32)
    return upgraded


def feature_column_names() -> list[str]:
    """ชื่อคอลัมน์สำหรับหัวตาราง CSV เพื่อให้เปิดดูข้อมูลดิบแล้วเข้าใจได้"""
    names: list[str] = []
    for slot in HAND_SLOTS:
        for i in range(NUM_LANDMARKS):
            names += [f"{slot}_{i}_x", f"{slot}_{i}_y", f"{slot}_{i}_z"]
        names.append(f"{slot}_present")
    names += [
        "Pair_offset_x",
        "Pair_offset_y",
        "Pair_offset_z",
        "Pair_scale_balance",
        "Pair_present",
    ]
    return names

File: app/js/features.js
/**
 * แปลงจุด landmark ของมือให้เป็นเวกเตอร์คุณลักษณะ 133 ค่า
 *
 * ไฟล์นี้เป็นฉบับ JavaScript ของ src/features.py และ **ต้องให้ผลตรงกันทุกหลัก**
 * เพราะข้อมูลที่เก็บจากมือถือกับจากคอมพิวเตอร์ต้องสลับไปมาได้
 * ถ้าแก้ไฟล์นี้ ต้องแก้ features.py ให้ตรงกันเสมอ แล้วรัน `run parity` เพื่อตรวจสอบ
 *
 * **คอนเวนชันกลาง:** เวกเตอร์ทุกตัวอ้างอิงกับ "ภาพดิบจากกล้อง (ยังไม่พลิกกระจก)"
 * ฝั่งนี้ส่งภาพดิบจาก <video> เข้า MediaPipe ตรง ๆ (การพลิกกระจกของกล้องหน้าทำด้วย
 * CSS transform ตอนแสดงผลเท่านั้น ไม่ได้แตะข้อมูลที่ส่งเข้าโมเดล) เวกเตอร์ที่ได้จึงอยู่
 * ในคอนเวนชันกลางอยู่แล้ว ไม่ต้องแปลงอะไร ส่วนฝั่งคอมพิวเตอร์ที่พลิกเฟรมก่อนตรวจจับ
 * ต้องแปลงกลับด้วย build_canonical_feature_vector (ดู src/features.py)
 *
 * ดูคำอธิบายว่าทำไมต้อง normalize ได้ที่ src/features.py และ docs/
 */

export const NUM_LANDMARKS = 21;
export const COORDS_PER_LANDMARK = 3;
export const PER_HAND_DIMS = NUM_LANDMARKS * COORDS_PER_LANDMARK + 1; // 64
export const HAND_SLOTS = ["Left", "Right"];
export const HANDS_DIMS = PER_HAND_DIMS * HAND_SLOTS.length; // 128

// ท่อนคู่มือ: การ normalize เทียบข้อมือของแต่ละมือเอง ทำให้ความสัมพันธ์ระหว่าง
// สองมือหายไปหมด (มือประกบกันกับมือแยกห่างกันได้ตัวเลขชุดเดียวกัน) อีก 5 ค่านี้
// เก็บส่วนที่หายไปกลับมา — ดูคำอธิบายเต็มที่ src/features.py
export const PAIR_DIMS = 5;
export const PAIR_START = HANDS_DIMS;
export const FEATURE_DIMS = HANDS_DIMS + PAIR_DIMS; // 133

/** จำนวนมิติของชุดข้อมูลรุ่นก่อนที่จะมีท่อนคู่มือ ใช้ตอนนำเข้าข้อมูลเก่า */
export const LEGACY_FEATURE_DIMS = HANDS_DIMS;

/**
 * normalize จุดของมือหนึ่งข้าง คืนอาเรย์ 63 ค่า
 * @param {Array<{x:number,y:number,z:number}>} landmarks จุด 21 จุดจาก MediaPipe
 * @returns {Float32Array}
 */
export function normalizeLandmarks(landmarks) {
  return normalizeWithGeometry(landmarks).normalized;
}

/**
 * เหมือน normalizeLandmarks แต่คืนตำแหน่งข้อมือและสเกลที่หารทิ้งไปด้วย
 * สองค่านี้คือสิ่งที่ท่อนคู่มือต้องใช้
 * @param {Array<{x:number,y:number,z:number}>} landmarks
 * @returns {{normalized: Float32Array, wrist: {x:number,y:number,z:number}, scale: number}}
 */
export function normalizeWithGeometry(landmarks) {
  const wrist = landmarks[0];

  // 1) ย้ายข้อมือไปที่จุดกำเนิด — ตัดผลของตำแหน่งในเฟรม
  const centered = new Float32Array(NUM_LANDMARKS * COORDS_PER_LANDMARK);
  for (let i = 0; i < NUM_LANDMARKS; i++) {
    const p = landmarks[i];
    centered[i * 3] = p.x - wrist.x;
    centered[i * 3 + 1] = p.y - wrist.y;
    centered[i * 3 + 2] = p.z - wrist.z;
  }

  // 2) หารด้วยระยะที่ไกลที่สุดจากข้อมือ — ตัดผลของขนาด/ระยะห่างจากกล้อง
  let scale = 0;
  for (let i = 0; i < NUM_LANDMARKS; i++) {
    const x = centered[i * 3];
    const y = centered[i * 3 + 1];
    const z = centered[i * 3 + 2];
    const distance = Math.hypot(x, y, z);
    if (distance > scale) scale = distance;
  }
  if (scale < 1e-6) scale = 1.0; // กันหารด้วยศูนย์

  for (let i = 0; i < centered.length; i++) centered[i] /= scale;

  // ปัดสเกลเป็น float32 ก่อนคืนออกไป เพราะฝั่ง Python เก็บค่านี้เป็น float32
  // ถ้าปล่อยเป็น float64 ท่อนคู่มือของสองฝั่งจะต่างกันเกินเกณฑ์ที่ `run parity` ยอมรับ
  return { normalized: centered, wrist, scale: Math.fround(scale) };
}

/**
 * รวมมือที่ตรวจเจอ (0-2 ข้าง) เป็นเวกเตอร์ความยาวคงที่ 133 ค่า
 *
 * โครงของเวกเตอร์:
 *   [  0..63 ]  มือ Left  : 63 ค่า normalize แล้ว + ธงบอกการมีอยู่
 *   [ 64..127]  มือ Right : เหมือนกัน
 *   [128..132]  ท่อนคู่มือ : offset x/y/z, สมดุลขนาด, ธงบอกว่าค่าใช้ได้
 *
 * @param {Array<{landmarks:Array, handedness:string}>} hands
 * @returns {Float32Array}
 */
export function buildFeatureVector(hands) {
  const vector = new Float32Array(FEATURE_DIMS);
  const wrists = new Array(HAND_SLOTS.length).fill(null);
  const scales = new Array(HAND_SLOTS.length).fill(0);

  for (const hand of hands) {
    const slot = HAND_SLOTS.indexOf(hand.handedness);
    if (slot < 0) continue;

    const start = slot * PER_HAND_DIMS;
    const { normalized, wrist, scale } = normalizeWithGeometry(hand.landmarks);
    vector.set(normalized, start);
    vector[start + PER_HAND_DIMS - 1] = 1.0; // ธง: มือข้างนี้ปรากฏอยู่
    wrists[slot] = wrist;
    scales[slot] = scale;
  }

  if (wrists.every((wrist) => wrist !== null)) fillPairBlock(vector, wrists, scales);

  return vector;
}

/**
 * เติมท่อนคู่มือ — เรียกเฉพาะเมื่อเจอมือครบสองข้างเท่านั้น
 *
 * offset หารด้วยขนาดมือเฉลี่ย เพื่อให้ยังไม่ขึ้นกับระยะห่างจากกล้อง
 * สมดุลขนาดใช้ (ขวา-ซ้าย)/(ขวา+ซ้าย) ซึ่งอยู่ในช่วง -1 ถึง 1 เสมอ
 * ดูเหตุผลของทั้งสองสูตรที่ src/features.py
 */
function fillPairBlock(vector, wrists, scales) {
  const [leftWrist, rightWrist] = wrists;
  const [leftScale, rightScale] = scales;

  let meanScale = Math.fround((leftScale + rightScale) / 2);
  if (meanScale < 1e-6) meanScale = 1.0;

  vector[PAIR_START] = (rightWrist.x - leftWrist.x) / meanScale;
  vector[PAIR_START + 1] = (rightWrist.y - leftWrist.y) / meanScale;
  vector[PAIR_START + 2] = (rightWrist.z - leftWrist.z) / meanScale;

  const totalScale = Math.fround(leftScale + rightScale);
  if (totalScale > 1e-6) {
    vector[PAIR_START + 3] = (rightScale - leftScale) / totalScale;
  }

  vector[PAIR_START + 4] = 1.0; // ธง: ค่าในท่อนนี้วัดมาจริง ไม่ใช่ศูนย์เพราะไม่รู้
}

/**
 * ขยายเวกเตอร์รุ่นเก่า 128 ค่า ให้เป็น 133 ค่า
 *
 * ท่อนคู่มือของข้อมูลเก่ากู้กลับไม่ได้จริง ๆ จึงเติมศูนย์แล้วปล่อยธงเป็น 0
 * ซึ่งหมายถึง "ไม่รู้ค่า" ไม่ใช่ "สองมือซ้อนทับกันพอดี" (ดู src/features.py)
 * @param {Float32Array|Array<number>} values
 * @returns {Float32Array}
 */
export function upgradeLegacyVector(values) {
  const upgraded = new Float32Array(FEATURE_DIMS);
  for (let i = 0; i < LEGACY_FEATURE_DIMS; i++) upgraded[i] = values[i];
  return upgraded;
}

/**
 * สร้างเวกเตอร์ของภาพสะท้อนกระจก ใช้เพิ่มข้อมูลตอนเทรน
 * กลับเครื่องหมายแกน x และสลับช่องมือซ้าย-ขวาพร้อมกัน
 * @param {Float32Array|Array<number>} vector
 * @returns {Float32Array}
 */
export function mirrorFeatureVector(vector) {
  const mirrored = new Float32Array(FEATURE_DIMS);

  for (let slot = 0; slot < HAND_SLOTS.length; slot++) {
    const start = slot * PER_HAND_DIMS;
    const present = vector[start + PER_HAND_DIMS - 1];
    const other = (slot + 1) % HAND_SLOTS.length;
    const dest = other * PER_HAND_DIMS;

    for (let i = 0; i < NUM_LANDMARKS; i++) {
      mirrored[dest + i * 3] = -vector[start + i * 3]; // กลับแกน x
      mirrored[dest + i * 3 + 1] = vector[start + i * 3 + 1];
      mirrored[dest + i * 3 + 2] = vector[start + i * 3 + 2];
    }
    mirrored[dest + PER_HAND_DIMS - 1] = present;
  }

  // ท่อนคู่มือกลับด้าน "ตรงข้าม" กับท่อนของแต่ละมือ: แกน x คงเดิม แกน y กับ z สลับ
  // เพราะการสลับช่องซ้าย-ขวากลับทิศของตัวลบเอง ซึ่งหักล้างกับการพลิกภาพบนแกน x พอดี
  // ดูการพิสูจน์ทีละบรรทัดที่ src/features.py
  mirrored[PAIR_START] = vector[PAIR_START];
  mirrored[PAIR_START + 1] = -vector[PAIR_START + 1];
  mirrored[PAIR_START + 2] = -vector[PAIR_START + 2];
  mirrored[PAIR_START + 3] = -vector[PAIR_START + 3];
  mirrored[PAIR_START + 4] = vector[PAIR_START + 4];

  return mirrored;
}

/** ชื่อคอลัมน์ ใช้ตอน export เป็น CSV ให้ตรงกับฝั่ง Python */
export function featureColumnNames() {
  const names = [];
  for (const slot of HAND_SLOTS) {
    for (let i = 0; i < NUM_LANDMARKS; i++) {
      names.push(`${slot}_${i}_x`, `${slot}_${i}_y`, `${slot}_${i}_z`);
    }
    names.push(`${slot}_present`);
  }
  names.push(
    "Pair_offset_x",
    "Pair_offset_y",
    "Pair_offset_z",
    "Pair_scale_balance",
    "Pair_present"
  );
  return names;
}
